### ISSUE TITLE
pretraining efficiency

### ISSUE THREAD

**Issue by pengzhangzhi** [uci-cbcl/esm-efficient#2]
Hi,

Amazing work!! I particularly interested in the pertaining efficiency in run time. A couple questions.

1. What is the major optimization for the improvement in training time? Is it flash attention? I'm assuming the checkpoint and other tricks might slow down the training for the memory. 
2. How do I use the esme for pertaining? Just load the esme model? Is the pretrained ckpt obtained from ESM2?

**Comment by MuhammedHasan** [comment]
Significant improvements are obtained using flash attention and partition attention. Additional memory optimizations do not significantly slow down the training process that much.

To train the model just create the model object and perform predictions:
```
model = ESM2(...)
logits = model(...)
loss = your_loss_fn(logits, ...)
``` 
see the arguments: https://github.com/uci-cbcl/esm-efficient/blob/116076d8c887642bb8f481c0d6e8b4325490d78f/esme/esm.py#L25

If you want to leverage partition attention properly, you need to prepare the batch structure for the input:
```
from esme.alphabet import tokenize_unpad
# tokenize without padding (more efficient avoids calculating with padding)
tokens, indices, cu_lens, max_len = tokenize_unpad(['MEEPQSDPSVEPPLSQETFSDLWK', 'MADQLTEEQIAEFKEAFSLFDKDG'])
tokens = tokens.to(0)
cu_lens = cu_lens.to(0)
log_probs = model.predict_log_prob(tokens, (cu_lens, max_len))
```

If you want to train the model for mask prediction objective, you can use the python ligthining class:
https://github.com/uci-cbcl/esm-efficient/blob/116076d8c887642bb8f481c0d6e8b4325490d78f/esme/trainer.py#L8

let me know if you have further questions.

**Comment by pengzhangzhi** [comment]
can we leverage partition attention in the training?

**Comment by MuhammedHasan** [comment]
Yes, partition attention should be used during the model training to achieve high GPU utilization, especially when you use DDP, so that you can create a batch of proteins with equal computational cost.

**Comment by pengzhangzhi** [comment]
Hii, thank you soo much!!!
How do I load a pretrained ESM2 model for example, ESM2-650M, the officially checkpoint?
```
# create load the model
model = ESM2.from_pretrained("{model}.safetensors", device=0)
```

**Comment by MuhammedHasan** [comment]
Weights are available here: https://huggingface.co/mhcelik/esm-efficient/tree/main

download it then just:
```
model = ESM2.from_pretrained("{model}.safetensors", device=0)
```
